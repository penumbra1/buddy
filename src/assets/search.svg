<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 90 90"><defs><style>.a{fill:none;}</style></defs><title>Icon by Alena Artemova from the Noun Project</title><circle class="a" cx="32.16" cy="25.41" r="21.05"/><path d="M56.19,56.62a6.15,6.15,0,0,1,2.31-1.09l-8.94-11.6a25.38,25.38,0,1,0-3.94,3l8.93,11.6A6.15,6.15,0,0,1,56.19,56.62ZM45,42.09a20.74,20.74,0,1,1,.08-.06Z"/><path d="M82.27,81.24l-18.52-24a5.08,5.08,0,0,0-5.28-1.7,6,6,0,0,0-4,3,5.11,5.11,0,0,0,.28,5.55l18.49,24a5.68,5.68,0,0,0,9-6.89Z"/></svg>