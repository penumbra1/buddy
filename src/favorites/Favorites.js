import React from "react";

const Favorites = () => {
  return null;
};

export default Favorites;
