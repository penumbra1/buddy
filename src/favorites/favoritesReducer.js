export const toggleFavorite = id => ({ type: "TOGGLE_FAVORITE", id });

export default (state = [], action) => {
  const { type, id } = action;
  if (type === "TOGGLE_FAVORITE" && id) {
    if (state.includes(id)) return state.filter(fav => fav !== id);
    else return [...state, id];
  } else return state;
};
