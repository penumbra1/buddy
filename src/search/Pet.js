import React, { lazy, Suspense } from "react";
import { Link } from "@reach/router";
import styled from "@emotion/styled";
import LocationIcon from "../assets/home.svg";
import colors from "../colors";

const Placeholder = lazy(() => import("../assets/placeholder.svg"));
const Home = styled(LocationIcon)`
  height: 1.5rem;
  margin-right: 0.5rem;
  fill: ${colors.greyDark};
`;

const StyledLink = styled(Link)`
  width: 100%;
  display: flex;
  margin: 0;
  background-color: ${colors.white};
  padding: 1.5rem;
  text-decoration: none;
  box-shadow: 1px 1px 1px 2px ${colors.shadow};

  img {
    width: 100%;
    height: 100%;
    object-fit: cover;
    object-position: center 40%;
  }

  .image-container {
    width: 100px;
    height: 100px;
    margin-right: 20px;
    border: 2px solid ${colors.greyLight};
    border-radius: 50%;
    overflow: hidden;
    flex-shrink: 0;
  }

  .info {
    height: 100px;
    display: flex;
    flex-direction: column;
    justify-content: space-around;
    overflow: hidden;
  }

  h1 {
    font-weight: normal;
    font-size: 2.5rem;
    width: 100%;
    margin: 0;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }

  h2 {
    font-weight: normal;
    font-size: 2rem;
    margin: -1rem 0 0 0;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }

  p {
    font-size: 1.7rem;
    margin: 0;
  }
`;

class Pet extends React.Component {
  render() {
    const { name, animal, breed, media, location, id } = this.props;
    let photos = [];
    if (media && media.photos && media.photos.photo) {
      photos = media.photos.photo.filter(photo => photo["@size"] === "pn");
    }

    let hero = null;
    if (photos[0] && photos[0].value) {
      hero = photos[0].value;
    }

    return (
      <StyledLink to={`/details/${id}`}>
        <div className="image-container">
          {hero ? (
            <img src={hero} alt={name} />
          ) : (
            <Suspense fallback={null}>
              <Placeholder style={{ fill: colors.greyDark }} />
            </Suspense>
          )}
        </div>
        <div className="info">
          <h1>{name}</h1>
          <h2>{breed}</h2>
          <p>
            <Home />
            {location}
          </p>
        </div>
      </StyledLink>
    );
  }
}

export default Pet;
