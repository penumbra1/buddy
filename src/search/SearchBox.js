import React from "react";
import { ANIMALS } from "petfinder-client";
import { search, getBreeds } from "../petfinder";
import Button from "../common/Button";
import styled from "@emotion/styled";

const StyledForm = styled.form`
  label {
    display: block;
    width: 60px;
  }

  input,
  select {
    margin-bottom: 30px;
    font-size: 18px;
    height: 30px;
    width: 320px;
  }
`;

// SearchBox is aware of the API but unaware of Redux.
// It manages transitory search parameter changes and data fetching.

class SearchBox extends React.Component {
  state = {
    ...this.props.initialParams,
    breedList: []
  };

  componentDidMount() {
    this.search();
    // Fetch breeds in without rerendering
    getBreeds(this.state.animal).then(breedList =>
      this.setState({ breedList })
    );
  }

  search = () => {
    const { breedList, ...searchParams } = this.state;
    const results = search(searchParams); // Promise
    this.props.onSearch(searchParams, results);
  };

  handleChange = event => {
    const { id, value } = event.target;
    if (id === "animal") {
      getBreeds(value).then(breedList => this.setState({ breedList }));
    }

    this.setState({ [id]: value });
  };

  handleFormSubmit = event => {
    event.preventDefault();
    this.search();
  };

  render() {
    return (
      <StyledForm onSubmit={this.handleFormSubmit}>
        <label htmlFor="location">
          Location
          <input
            id="location"
            value={this.state.location}
            onChange={this.handleChange}
            placeholder="Location"
          />
        </label>
        <label htmlFor="animal">
          Animal
          <select
            id="animal"
            value={this.state.animal}
            onChange={this.handleChange}
            onBlur={this.handleChange} // Remove it?
          >
            <option />
            {ANIMALS.map(animal => (
              <option key={animal} value={animal}>
                {animal}
              </option>
            ))}
          </select>
        </label>
        <label htmlFor="breed">
          Breed
          <select
            disabled={!this.state.breedList.length}
            id="breed"
            value={this.state.breed}
            onChange={this.handleChange}
            onBlur={this.handleChange}
          >
            <option />
            {this.state.breedList.map(breed => (
              <option key={breed} value={breed}>
                {breed}
              </option>
            ))}
          </select>
        </label>
        <Button>Submit</Button>
      </StyledForm>
    );
  }
}

export default SearchBox;
