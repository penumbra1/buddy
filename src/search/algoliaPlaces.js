// import debounce from "lodash.debounce";
import algolia from "algoliasearch/lite";

const client = algolia.initPlaces(
  process.env.ALGOLIA_ID,
  process.env.ALGOLIA_KEY
);

const searchByPrefix = prefix => {
  const params = {
    query: prefix,
    type: "city",
    countries: ["us"],
    language: "en"
  };
  return client
    .search(params)
    .then(res => res.hits)
    .catch(e => {
      console.error(e);
      return [];
    });
};

// const searchByPrefixDeb = debounce(searchByPrefix, 500);

const getCities = prefix => {
  if (prefix && prefix.length > 3) {
    return searchByPrefix(prefix).then(hits => {
      return hits.map(({ administrative: [state], locale_names: [name] }) => ({
        value: `${name}, ${state}`,
        label: `${name}, ${state}`
      }));
    });
  } else return Promise.resolve([]);
};

export default getCities;
