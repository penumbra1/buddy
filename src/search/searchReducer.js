export const updateSearch = changes => {
  return {
    type: "UPDATE_SEARCH",
    changes
  };
};

const defaultState = {
  location: "Seattle, WA",
  animal: "cat",
  breed: ""
};

export default (state = defaultState, action) => {
  const { type, changes } = action;
  if (type === "UPDATE_SEARCH") {
    if (changes.animal) changes.breed = "";
    return { ...state, ...changes };
  } else return state;
};
