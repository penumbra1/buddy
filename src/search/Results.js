import React from "react";
import styled from "@emotion/styled";
import Pet from "./Pet";

const Wrapper = styled.div`
  display: grid;
  grid-gap: 3rem 2.4rem;
  grid-template-columns: repeat(auto-fill, minmax(360px, 1fr));
`;

const Results = ({ pets }) => {
  return (
    <Wrapper>
      {pets.map(pet => {
        let breed;
        if (Array.isArray(pet.breeds.breed)) {
          breed = pet.breeds.breed.join(", ");
        } else {
          breed = pet.breeds.breed;
        }
        return (
          <Pet
            animal={pet.animal}
            key={pet.id}
            name={pet.name}
            breed={breed}
            media={pet.media}
            location={`${pet.contact.city}, ${pet.contact.state}`}
            id={pet.id}
          />
        );
      })}
    </Wrapper>
  );
};

export default Results;
