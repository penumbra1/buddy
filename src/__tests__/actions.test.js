import { updateSearch } from "../search/searchReducer";
import { toggleFavorite } from "../favorites/favoritesReducer";

test("updateSearch should create an action correctly", () => {
  const action = updateSearch({ location: "Cleveland, OH" });
  expect(action).toMatchSnapshot();
});

test("toggleFavorite should create an action correctly", () => {
  const action = toggleFavorite("1");
  expect(action).toMatchSnapshot();
});
