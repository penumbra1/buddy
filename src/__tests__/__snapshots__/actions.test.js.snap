// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toggleFavorite should create an action correctly 1`] = `
Object {
  "id": "1",
  "type": "TOGGLE_FAVORITE",
}
`;

exports[`updateSearch should create an action correctly 1`] = `
Object {
  "changes": Object {
    "location": "Cleveland, OH",
  },
  "type": "UPDATE_SEARCH",
}
`;
